### ISSUE TITLE
[no-unnecessary-type-assertion] false positive on templated values

### ISSUE THREAD

**Issue by nevir** [typescript-eslint/typescript-eslint#1410]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

```TS
function genericGet<TValue>(object: any, path: string): TValue {
  /* pretend this works. */
  return (undefined as any) as TValue;
}

const object = { foo: { bar: 123 } };
const result = genericGet(object, 'foo.bar') as number;
```

**Expected Result**

No error

**Actual Result**

```
const result = genericGet(object, 'foo.bar') as number;
//             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
//This assertion is unnecessary since it does not change the type of the expression.
```

**Versions**

Reproduces on `master`

**Comment by bradzacher** [comment]
This is due to weirdness in the typescript types API, illustrated by:
![image](https://user-images.githubusercontent.com/7462525/71927704-ec4c8480-314a-11ea-8a90-d442b0e93528.png)

Typescript notices that there's no type for the generic, and that there's an explicit cast on the return type, so it sets the type of the generic based on that.

Why do you have a method with a generic that you're not using?
Is this representative of your actual code? If so, it looks like this is terribly unsafe code.

**Comment by nevir** [comment]
Yeah, this is a bit of a contrived example. Where I'm running into it in the real world is with [`lodash.get`](https://lodash.com/docs/#get), which boils down to:

```ts
interface LoDashStatic {
  get<TObject, TResult>(
      object: TObject,
      path: StringRepresentable|StringRepresentable[],
      defaultValue?: TResult
  ): TResult;
}
```

```ts
const result = _.get(object, 'foo.bar') as number;
```

**Comment by nevir** [comment]
Without the cast, `result` is typed `unknown`.

However, once loaded, the checker keeps the inferred type around (e.g. [`originalExpression`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts#L271) is considered to be type `number`)

**Comment by bradzacher** [comment]
It's similar to the problem we ran into here: https://github.com/microsoft/TypeScript/issues/35431 where an assertion changes the type of the original statement, instead of just changing the new assertion statement.

**Comment by bradzacher** [comment]
Reading into it a bit more, I found https://github.com/microsoft/TypeScript/issues/31435, which means this is entirely intentional.

We might have to look into adding some rather complex checks to figure out if the thing being cast is a call expression with a signature that has a generic in the return type position...

A workaround for this is to instead use a type annotation:
```ts
const result: number = lodashGet(object, 'foo.bar');
```
Typescript uses the same logic to contextually infer the generic, but it won't trigger the rule.

**Comment by bradzacher** [comment]
As per  microsoft/TypeScript#36062, there is nothing we can do about this. This is intentional design by typescript, and as such the API will not give us any different data.
